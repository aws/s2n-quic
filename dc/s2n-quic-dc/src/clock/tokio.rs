// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use tokio::time::{self, sleep_until, Instant};

impl_clock!();

#[cfg(test)]
mod tests {
    use crate::clock::{tokio::Clock, Timer};
    use core::time::Duration;
    use s2n_quic_core::time::{clock::Timer as _, Clock as _};

    #[tokio::test]
    async fn clock_test() {
        let clock = Clock::default();
        let mut timer = Timer::new(&clock);
        timer.ready().await;
        timer.update(clock.get_time() + Duration::from_secs(1));
        timer.ready().await;
    }
}
