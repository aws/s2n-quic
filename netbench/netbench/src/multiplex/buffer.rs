// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use super::Frame;
use crate::Result;
use bytes::{Buf, BufMut, Bytes, BytesMut};
use core::{
    mem::MaybeUninit,
    task::{Context, Poll},
};
use futures::ready;
use std::{collections::VecDeque, io::IoSlice};
use tokio::io::ReadBuf;

const CHUNK_CAPACITY: usize = 4096;

#[derive(Debug, Default)]
pub struct ReadBuffer {
    len: usize,
    tail: BytesMut,
    head: VecDeque<BytesMut>,
}

impl ReadBuffer {
    pub fn read<F: FnOnce(&mut ReadBuf) -> Poll<Result<()>>>(
        &mut self,
        read: F,
    ) -> Poll<Result<()>> {
        let capacity = self.tail.capacity();

        if capacity == 0 {
            self.tail.reserve(CHUNK_CAPACITY);
        } else if capacity < 32 {
            let tail = core::mem::replace(&mut self.tail, BytesMut::with_capacity(CHUNK_CAPACITY));
            self.head.push_back(tail);
        }

        let buf = self.tail.chunk_mut();
        let buf = unsafe { &mut *(buf as *mut _ as *mut [MaybeUninit<u8>]) };
        let mut buf = ReadBuf::uninit(buf);
        ready!(read(&mut buf))?;
        let len = buf.filled().len();

        if len == 0 {
            return Ok(()).into();
        }

        unsafe {
            self.tail.advance_mut(len);
            self.len += len;
        }

        Ok(()).into()
    }

    fn check_consistency(&self) {
        if cfg!(debug_assertions) {
            let mut actual_len = self.tail.len();
            for chunk in self.head.iter() {
                actual_len += chunk.len();
            }
            assert_eq!(actual_len, self.remaining(), "{self:?}");
        }
    }
}

impl Buf for ReadBuffer {
    fn remaining(&self) -> usize {
        self.len
    }

    fn chunk(&self) -> &[u8] {
        for chunk in self.head.iter() {
            if !chunk.is_empty() {
                return chunk.chunk();
            }
        }

        self.tail.chunk()
    }

    fn advance(&mut self, mut cnt: usize) {
        while cnt > 0 {
            let len = if let Some(front) = self.head.front_mut() {
                let len = front.len().min(cnt);
                front.advance(len);
                if front.is_empty() {
                    let _ = self.head.pop_front();
                }
                len
            } else {
                self.tail.advance(cnt);
                cnt
            };

            cnt -= len;
            self.len -= len;
        }

        self.check_consistency();
    }

    fn copy_to_bytes(&mut self, len: usize) -> Bytes {
        while let Some(mut front) = self.head.pop_front() {
            if front.is_empty() {
                continue;
            }

            self.len -= len;

            if front.len() == len {
                self.check_consistency();
                return front.freeze();
            }

            let out = front.split_to(len);
            self.head.push_front(front);

            self.check_consistency();
            return out.freeze();
        }

        let out = self.tail.split_to(len).freeze();

        self.len -= len;
        self.check_consistency();

        out
    }
}

#[derive(Debug, Default)]
pub struct WriteBuffer {
    unfilled: BytesMut,
    queue: VecDeque<Bytes>,
    push_interest: bool,
}

impl WriteBuffer {
    pub fn request_push(&mut self, max_queue_len: usize) -> bool {
        let can_push = max_queue_len > self.queue.len();
        self.push_interest |= !can_push;
        can_push
    }

    pub fn push(&mut self, frame: Frame) {
        let capacity = self.unfilled.capacity();

        if capacity == 0 {
            self.unfilled.reserve(CHUNK_CAPACITY);
        } else if capacity < 32 {
            self.unfilled = BytesMut::with_capacity(CHUNK_CAPACITY);
        }

        frame.write_header(&mut self.unfilled);
        self.queue.push_back(self.unfilled.split().freeze());

        if let Some(data) = frame.body() {
            self.queue.push_back(data);
        }
    }

    pub fn push_priority(&mut self, frame: Frame) {
        let capacity = self.unfilled.capacity();

        const CAPACITY: usize = 4096;

        if capacity == 0 {
            self.unfilled.reserve(CAPACITY);
        } else if capacity < 32 {
            self.unfilled = BytesMut::with_capacity(CAPACITY);
        }

        frame.write_header(&mut self.unfilled);
        let header = self.unfilled.split().freeze();

        // push the body first
        if let Some(data) = frame.body() {
            self.queue.push_front(data);
        }

        self.queue.push_front(header);
    }

    pub fn is_empty(&self) -> bool {
        self.queue.is_empty()
    }

    pub fn pop_front(&mut self) -> Option<Bytes> {
        self.queue.pop_front()
    }

    pub fn push_front(&mut self, chunk: Bytes) {
        self.queue.push_front(chunk);
    }

    pub fn notify(&mut self, cx: &mut Context) {
        if self.push_interest {
            self.push_interest = false;
            cx.waker().wake_by_ref();
        }
    }

    pub fn advance(&mut self, mut len: usize) {
        while let Some(mut chunk) = self.queue.pop_front() {
            let next_len = len.saturating_sub(chunk.len());

            if next_len > 0 {
                len = next_len;
                continue;
            }

            if chunk.len() > len {
                chunk.advance(len);
                self.queue.push_front(chunk);
            }

            return;
        }
    }

    pub fn chunks(&self) -> Vec<IoSlice> {
        self.queue.iter().map(|v| IoSlice::new(v)).collect()
    }
}
