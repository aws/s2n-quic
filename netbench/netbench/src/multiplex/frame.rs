// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use crate::{connection::Owner, Result};
use bytes::{Buf, BufMut, Bytes};
use core::mem::size_of;
use enum_primitive_derive::Primitive;
use num_traits::FromPrimitive;
use std::io;

#[derive(Clone, Copy, Debug, Primitive)]
#[repr(u8)]
enum Tag {
    StreamOpen = 0,
    StreamData = 1,
    StreamFinish = 2,
    MaxStreams = 3,
    MaxStreamData = 4,
    InitialMaxStreamData = 5,
}

#[derive(Clone, Debug)]
pub enum Frame {
    StreamOpen { id: u64, bidirectional: bool },
    StreamData { id: u64, owner: Owner, data: Bytes },
    StreamFinish { id: u64, owner: Owner },
    MaxStreams { up_to: u64 },
    MaxStreamData { id: u64, owner: Owner, up_to: u64 },
    InitialMaxStreamData { up_to: u64 },
}

impl Frame {
    pub fn write_header<B: BufMut>(&self, buf: &mut B) {
        match self {
            Self::StreamOpen { id, bidirectional } => {
                buf.put_u8(Tag::StreamOpen as _);
                buf.put_u64(*id);
                buf.put_u8(*bidirectional as u8);
            }
            Self::StreamData { id, owner, data } => {
                buf.put_u8(Tag::StreamData as _);
                buf.put_u64(*id);
                buf.put_u8(*owner as _);
                buf.put_u32(data.len() as _);
            }
            Self::StreamFinish { id, owner } => {
                buf.put_u8(Tag::StreamFinish as _);
                buf.put_u64(*id);
                buf.put_u8(*owner as _);
            }
            Self::MaxStreams { up_to } => {
                buf.put_u8(Tag::MaxStreams as _);
                buf.put_u64(*up_to);
            }
            Self::MaxStreamData { id, owner, up_to } => {
                buf.put_u8(Tag::MaxStreamData as _);
                buf.put_u64(*id);
                buf.put_u8(*owner as _);
                buf.put_u64(*up_to);
            }
            Self::InitialMaxStreamData { up_to } => {
                buf.put_u8(Tag::InitialMaxStreamData as _);
                buf.put_u64(*up_to);
            }
        }
    }

    pub fn body(self) -> Option<Bytes> {
        if let Self::StreamData { data, .. } = self {
            Some(data)
        } else {
            None
        }
    }
}

#[derive(Debug, Default)]
pub struct Decoder {
    tag: Option<Tag>,
    stream: Option<(u64, Owner, usize)>,
}

impl Decoder {
    pub fn decode<B: Buf>(&mut self, buf: &mut B) -> Result<Option<Frame>> {
        if let Some((id, owner, len)) = self.stream.take() {
            return self.decode_stream(buf, id, owner, len);
        }

        if let Some(tag) = self.tag.take() {
            return self.decode_frame(buf, tag);
        }

        if buf.remaining() == 0 {
            return Ok(None);
        }

        let tag = Tag::from_u8(buf.get_u8())
            .ok_or_else(|| io::Error::new(io::ErrorKind::Other, "invalid frame tag"))?;

        self.decode_frame(buf, tag)
    }

    fn decode_frame<B: Buf>(&mut self, buf: &mut B, tag: Tag) -> Result<Option<Frame>> {
        match tag {
            Tag::StreamOpen => {
                if buf.remaining() < (size_of::<u64>() + size_of::<u8>()) {
                    self.tag = Some(tag);
                    return Ok(None);
                }

                let id = buf.get_u64();
                let bidirectional = buf.get_u8() != 0;

                Ok(Some(Frame::StreamOpen { id, bidirectional }))
            }
            Tag::StreamData => {
                if buf.remaining() < (size_of::<u64>() + size_of::<u8>() + size_of::<u32>()) {
                    self.tag = Some(tag);
                    return Ok(None);
                }

                let id = buf.get_u64();
                let owner = Self::decode_owner(buf)?;
                let len = buf.get_u32();
                self.decode_stream(buf, id, owner, len as _)
            }
            Tag::StreamFinish => {
                if buf.remaining() < (size_of::<u64>() + size_of::<u8>()) {
                    self.tag = Some(tag);
                    return Ok(None);
                }

                let id = buf.get_u64();
                let owner = Self::decode_owner(buf)?;

                Ok(Some(Frame::StreamFinish { id, owner }))
            }
            Tag::MaxStreams => {
                if buf.remaining() < size_of::<u64>() {
                    self.tag = Some(tag);
                    return Ok(None);
                }

                let up_to = buf.get_u64();

                Ok(Some(Frame::MaxStreams { up_to }))
            }
            Tag::MaxStreamData => {
                if buf.remaining() < (size_of::<u64>() + size_of::<u8>() + size_of::<u64>()) {
                    self.tag = Some(tag);
                    return Ok(None);
                }

                let id = buf.get_u64();
                let owner = Self::decode_owner(buf)?;
                let up_to = buf.get_u64();

                Ok(Some(Frame::MaxStreamData { id, owner, up_to }))
            }
            Tag::InitialMaxStreamData => {
                if buf.remaining() < (size_of::<u64>()) {
                    self.tag = Some(tag);
                    return Ok(None);
                }

                let up_to = buf.get_u64();

                Ok(Some(Frame::InitialMaxStreamData { up_to }))
            }
        }
    }

    fn decode_owner<B: Buf>(buf: &mut B) -> Result<Owner> {
        let owner = buf.get_u8();
        Ok(match owner {
            0 => Owner::Local,
            1 => Owner::Remote,
            _ => return Err(io::Error::new(io::ErrorKind::InvalidData, "invalid owner id").into()),
        })
    }

    fn decode_stream<B: Buf>(
        &mut self,
        buf: &mut B,
        id: u64,
        owner: Owner,
        len: usize,
    ) -> Result<Option<Frame>> {
        if len == 0 {
            return self.decode(buf);
        }

        let chunk_len = buf.chunk().len();

        if chunk_len == 0 {
            self.stream = Some((id, owner, len));
            return Ok(None);
        }

        Ok(if chunk_len >= len {
            let data = buf.copy_to_bytes(len);
            Some(Frame::StreamData { id, owner, data })
        } else {
            let data = buf.copy_to_bytes(chunk_len);
            self.stream = Some((id, owner, len - chunk_len));
            Some(Frame::StreamData { id, owner, data })
        })
    }
}
