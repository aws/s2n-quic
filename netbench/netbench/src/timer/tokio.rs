// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use super::Timestamp;
use core::{
    future::Future,
    pin::Pin,
    task::{Context, Poll},
    time::Duration,
};
use futures::ready;
use tokio::time::{sleep, Instant, Sleep};

#[derive(Debug)]
pub struct Timer {
    start: Instant,
    target: Option<Timestamp>,
    sleep: Pin<Box<Sleep>>,
}

impl Timer {
    fn set_target(&mut self, target: Timestamp) {
        self.target = Some(target);
        let duration = unsafe { target.as_duration() };
        self.sleep.as_mut().reset(self.start + duration);
    }
}

impl Default for Timer {
    fn default() -> Self {
        Self {
            start: Instant::now(),
            target: None,
            sleep: Box::pin(sleep(Duration::from_secs(0))),
        }
    }
}

impl super::Timer for Timer {
    fn now(&self) -> Timestamp {
        let duration = self.start.elapsed();
        unsafe { Timestamp::from_duration(duration) }
    }

    fn poll(&mut self, target: Timestamp, cx: &mut Context) -> Poll<()> {
        if let Some(prev_target) = self.target.as_mut() {
            if *prev_target != target {
                self.set_target(target);
            }
        } else {
            self.set_target(target);
        }

        ready!(self.sleep.as_mut().poll(cx));

        self.target = None;

        Poll::Ready(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::timer::Timer as _;
    use futures_test::task::new_count_waker;

    #[tokio::test(start_paused = true)]
    async fn timer_test() {
        let mut timer = Timer::default();
        let (waker, _count) = new_count_waker();
        let mut cx = Context::from_waker(&waker);

        tokio::time::advance(Duration::from_secs(1)).await;

        let mut now = timer.now();

        let mut times = [now; 5];
        for (idx, time) in times.iter_mut().enumerate() {
            *time += Duration::from_secs(idx as _);
        }

        assert!(timer.poll(now, &mut cx).is_ready());

        for _ in 0..times.len() {
            // poll a bunch of different times in loop to make sure all of the branches are covered
            for time in times.iter().chain(times.iter().rev()).copied() {
                assert_eq!(timer.poll(time, &mut cx).is_ready(), time <= now);
                assert_eq!(timer.poll(time, &mut cx).is_ready(), time <= now);
            }

            tokio::time::advance(Duration::from_secs(1)).await;
            now += Duration::from_secs(1);
        }
    }
}
