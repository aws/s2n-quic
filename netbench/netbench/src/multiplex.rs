// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use crate::{connection::Owner, Result};
use bytes::Buf;
use core::{
    pin::Pin,
    task::{Context, Poll},
};
use futures::ready;
use std::collections::{hash_map::Entry, HashMap, HashSet, VecDeque};
use tokio::io::{AsyncRead, AsyncWrite};

mod buffer;
mod frame;
mod stream;

use buffer::{ReadBuffer, WriteBuffer};
use frame::Frame;
use stream::{ReceiveStream, SendStream, Stream};

const DEFAULT_STREAM_WINDOW: u64 = 256000;
const DEFAULT_MAX_STREAMS: u64 = 100;

#[derive(Clone, Debug)]
pub struct Config {
    pub stream_window: u64,
    pub max_streams: u64,
    pub max_stream_data_thresh: u64,
    pub max_stream_frame_len: u32,
    pub max_write_queue_len: usize,
    pub peer_max_streams: u64,
    pub peer_max_stream_data: u64,
}

impl Default for Config {
    fn default() -> Self {
        Self {
            stream_window: DEFAULT_STREAM_WINDOW,
            max_streams: DEFAULT_MAX_STREAMS,
            max_stream_data_thresh: DEFAULT_STREAM_WINDOW / 2,
            max_stream_frame_len: (1 << 15),
            max_write_queue_len: 250,
            peer_max_streams: DEFAULT_MAX_STREAMS,
            peer_max_stream_data: DEFAULT_STREAM_WINDOW,
        }
    }
}

#[derive(Debug)]
pub struct Connection<T: AsyncRead + AsyncWrite> {
    id: u64,
    inner: Pin<Box<T>>,
    rx_open: bool,
    tx_open: bool,
    frame: Option<frame::Frame>,
    read_buf: ReadBuffer,
    decoder: frame::Decoder,
    write_buf: WriteBuffer,
    stream_controller: stream::Controller,
    streams: [HashMap<u64, Stream>; 2],
    max_stream_data: HashSet<(Owner, u64)>,
    pending_accept: VecDeque<u64>,
    config: Config,
}

impl<T: AsyncRead + AsyncWrite> Connection<T> {
    pub fn new(id: u64, inner: Pin<Box<T>>, config: Config) -> Self {
        let mut write_buf = WriteBuffer::default();

        if config.stream_window != DEFAULT_STREAM_WINDOW {
            write_buf.push(Frame::InitialMaxStreamData {
                up_to: config.stream_window,
            });
        }

        if config.max_streams != DEFAULT_MAX_STREAMS {
            write_buf.push(Frame::MaxStreams {
                up_to: config.max_streams,
            });
        }

        Self {
            id,
            inner,
            rx_open: true,
            tx_open: true,
            frame: None,
            read_buf: Default::default(),
            decoder: Default::default(),
            write_buf,
            stream_controller: stream::Controller::new(config.max_streams, config.peer_max_streams),
            streams: [Default::default(), Default::default()],
            max_stream_data: Default::default(),
            pending_accept: Default::default(),
            config,
        }
    }

    fn flush_write_buffer(&mut self, cx: &mut Context) -> Result<()> {
        if !self.tx_open {
            return if self.write_buf.is_empty() {
                Ok(())
            } else {
                Err("stream was closed with pending data".into())
            };
        }

        if self.write_buf.is_empty() {
            return Ok(());
        }

        if self.inner.as_ref().is_write_vectored() {
            let chunks = self.write_buf.chunks();

            return match self.inner.as_mut().poll_write_vectored(cx, &chunks) {
                Poll::Ready(result) => {
                    let len = result?;
                    self.write_buf.advance(len);
                    self.write_buf.notify(cx);
                    Ok(())
                }
                Poll::Pending => Ok(()),
            };
        }

        while let Some(mut chunk) = self.write_buf.pop_front() {
            match self.inner.as_mut().poll_write(cx, &chunk) {
                Poll::Ready(result) => {
                    let len = result?;
                    chunk.advance(len);

                    if !chunk.is_empty() {
                        self.write_buf.push_front(chunk);
                    }

                    if len == 0 {
                        return if self.write_buf.is_empty() {
                            self.tx_open = false;
                            Ok(())
                        } else {
                            Err("stream was closed with pending data".into())
                        };
                    }
                }
                Poll::Pending => {
                    self.write_buf.push_front(chunk);
                    break;
                }
            }
        }

        self.write_buf.notify(cx);

        Ok(())
    }

    fn flush_read_buffer(&mut self, cx: &mut Context) -> Result<()> {
        loop {
            if self.frame.is_none() {
                if let Some(frame) = self.decoder.decode(&mut self.read_buf)? {
                    self.frame = Some(frame);
                }
            }

            match self.dispatch_frame(cx) {
                Poll::Ready(result) => result?,
                Poll::Pending => return Ok(()),
            }
        }
    }

    fn fill_read_buffer(&mut self, cx: &mut Context) -> Poll<Result<()>> {
        // don't fill the buffer if we have a pending frame
        if self.frame.is_some() {
            return Poll::Pending;
        }

        // the socket is closed
        if !self.rx_open {
            return Ok(()).into();
        }

        let rx_open = &mut self.rx_open;
        let inner = self.inner.as_mut();

        ready!(self.read_buf.read(|buf| {
            ready!(inner.poll_read(cx, buf))?;

            // the socket returned a 0 write
            if buf.filled().is_empty() {
                *rx_open = false;
            }

            Ok(()).into()
        }))?;

        Ok(()).into()
    }

    fn dispatch_frame(&mut self, cx: &mut Context) -> Poll<Result<()>> {
        use frame::Frame::*;

        match self.frame.take() {
            Some(StreamOpen { id, bidirectional }) => {
                // TODO make sure the peer hasn't opened too many
                let mut stream = Stream {
                    rx: Some(ReceiveStream::new(self.config.stream_window)),
                    tx: None,
                };
                if bidirectional {
                    stream.tx = Some(SendStream::new(self.config.peer_max_stream_data));
                };
                self.streams[Owner::Remote].insert(id, stream);
                self.pending_accept.push_back(id);
                cx.waker().wake_by_ref();
            }
            Some(StreamData {
                id,
                owner,
                mut data,
            }) => {
                if let Some(rx) = self.streams[owner]
                    .get_mut(&id)
                    .and_then(|stream| stream.rx.as_mut())
                {
                    let len = data.len() as u64;
                    let len = rx.buffer(len)?;
                    data.advance(len as _);

                    if !data.is_empty() {
                        self.frame = Some(StreamData { id, owner, data });
                        return Poll::Pending;
                    }
                }
            }
            Some(MaxStreams { up_to }) => {
                self.stream_controller.max_streams(up_to);
            }
            Some(MaxStreamData { id, owner, up_to }) => {
                if let Some(tx) = self.streams[owner]
                    .get_mut(&id)
                    .and_then(|stream| stream.tx.as_mut())
                {
                    tx.max_data(up_to);
                }
            }
            Some(StreamFinish { id, owner }) => {
                if let Some(rx) = self.streams[owner]
                    .get_mut(&id)
                    .ok_or("invalid stream")?
                    .rx
                    .as_mut()
                {
                    rx.finish();
                }
            }
            Some(InitialMaxStreamData { up_to }) => {
                self.config.peer_max_stream_data = up_to;
            }
            None => return Poll::Pending,
        }

        Ok(()).into()
    }
}

impl<T: AsyncRead + AsyncWrite> super::Connection for Connection<T> {
    fn id(&self) -> u64 {
        self.id
    }

    fn poll_open_bidirectional_stream(&mut self, id: u64, cx: &mut Context) -> Poll<Result<()>> {
        ready!(self.stream_controller.open(cx));

        self.write_buf.push(Frame::StreamOpen {
            id,
            bidirectional: true,
        });

        let stream = Stream {
            rx: Some(ReceiveStream::new(self.config.stream_window)),
            tx: Some(SendStream::new(self.config.peer_max_stream_data)),
        };
        self.streams[Owner::Local].insert(id, stream);

        Ok(()).into()
    }

    fn poll_open_send_stream(&mut self, id: u64, cx: &mut Context) -> Poll<Result<()>> {
        ready!(self.stream_controller.open(cx));

        self.write_buf.push(Frame::StreamOpen {
            id,
            bidirectional: false,
        });

        let stream = Stream {
            rx: None,
            tx: Some(SendStream::new(self.config.peer_max_stream_data)),
        };
        self.streams[Owner::Local].insert(id, stream);

        Ok(()).into()
    }

    fn poll_accept_stream(&mut self, _cx: &mut Context) -> Poll<Result<Option<u64>>> {
        if !self.rx_open {
            return Ok(None).into();
        }

        if let Some(id) = self.pending_accept.pop_front() {
            Ok(Some(id)).into()
        } else {
            Poll::Pending
        }
    }

    fn poll_send(
        &mut self,
        owner: Owner,
        id: u64,
        bytes: u64,
        cx: &mut Context,
    ) -> Poll<Result<u64>> {
        if !self.write_buf.request_push(self.config.max_write_queue_len) {
            return Poll::Pending;
        }

        let stream = self.streams[owner]
            .get_mut(&id)
            .ok_or("missing stream")?
            .tx
            .as_mut()
            .ok_or("missing tx stream")?;

        let allowed_bytes = bytes.min(self.config.max_stream_frame_len as _);

        if let Some(data) = stream.send(allowed_bytes, cx) {
            let len = data.len() as u64;
            self.write_buf.push(frame::Frame::StreamData {
                id,
                owner: !owner,
                data,
            });
            Ok(len).into()
        } else {
            Poll::Pending
        }
    }

    fn poll_receive(
        &mut self,
        owner: Owner,
        id: u64,
        bytes: u64,
        cx: &mut Context,
    ) -> Poll<Result<u64>> {
        let stream = self.streams[owner]
            .get_mut(&id)
            .ok_or("missing stream")?
            .rx
            .as_mut()
            .ok_or("missing rx stream")?;

        let len = ready!(stream.receive(bytes, cx))?;

        if stream.receive_window() < self.config.stream_window / 2 {
            self.max_stream_data.insert((owner, id));
        }

        Ok(len).into()
    }

    fn poll_send_finish(&mut self, owner: Owner, id: u64, _cx: &mut Context) -> Poll<Result<()>> {
        if !self.write_buf.request_push(self.config.max_write_queue_len) {
            return Poll::Pending;
        }

        if let Entry::Occupied(mut entry) = self.streams[owner].entry(id) {
            let stream = entry.get_mut();

            if stream.tx.take().is_some() {
                self.write_buf
                    .push(frame::Frame::StreamFinish { id, owner: !owner });
            }

            if stream.rx.is_none() {
                entry.remove();
                self.stream_controller.close();
            }
        }

        Ok(()).into()
    }

    fn poll_receive_finish(&mut self, owner: Owner, id: u64, cx: &mut Context) -> Poll<Result<()>> {
        if let Entry::Occupied(mut entry) = self.streams[owner].entry(id) {
            let stream = entry.get_mut();

            if let Some(rx) = stream.rx.as_mut() {
                ready!(rx.poll_finish(cx));
            }
            stream.rx = None;

            if stream.tx.is_none() {
                entry.remove();
                self.stream_controller.close();
            }
        }

        Ok(()).into()
    }

    fn poll_progress(&mut self, cx: &mut Context) -> Poll<Result<()>> {
        loop {
            if let Some(up_to) = self.stream_controller.transmit() {
                self.write_buf.push_priority(Frame::MaxStreams { up_to });
            }

            for (owner, id) in self.max_stream_data.drain() {
                if let Some(stream) = self.streams[owner].get_mut(&id).and_then(|s| s.rx.as_mut()) {
                    let up_to = stream.credit(self.config.stream_window);
                    self.write_buf.push_priority(Frame::MaxStreamData {
                        id,
                        owner: !owner,
                        up_to,
                    });
                }
            }

            self.flush_write_buffer(cx)?;
            self.flush_read_buffer(cx)?;
            ready!(self.fill_read_buffer(cx))?;

            // the connection is done
            if !self.rx_open && self.frame.is_none() {
                return Ok(()).into();
            }
        }
    }

    fn poll_finish(&mut self, cx: &mut Context) -> Poll<Result<()>> {
        if self.tx_open {
            self.flush_write_buffer(cx)?;

            // wait to shutdown the socket until we have written everything
            if !self.write_buf.is_empty() {
                return Poll::Pending;
            }

            // notify the peer we're not writing anything anymore
            ready!(self.inner.as_mut().poll_shutdown(cx))?;
            self.tx_open = false;
        }

        loop {
            // work to read all of the remaining data
            self.flush_read_buffer(cx)?;
            ready!(self.fill_read_buffer(cx))?;

            // the connection is done
            if !self.rx_open && self.frame.is_none() {
                return Ok(()).into();
            }
        }
    }
}

#[derive(Debug, Default)]
struct Blocked(Option<core::task::Waker>);

impl Blocked {
    pub fn block(&mut self, cx: &mut Context) {
        self.0 = Some(cx.waker().clone());
    }

    pub fn unblock(&mut self) {
        if let Some(waker) = self.0.take() {
            waker.wake();
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{scenario::Scenario, testing, timer, trace::MemoryLogger, units::*, Driver};
    use futures_test::task::new_count_waker;
    use insta::assert_display_snapshot;
    use std::collections::HashSet;

    fn test(config: Config, scenario: &Scenario) -> (MemoryLogger, MemoryLogger) {
        let traces = &scenario.traces;

        let (client, server) = testing::Connection::pair(10000);

        let mut client = {
            let scenario = &scenario.clients[0].connections[0];
            let conn = Box::pin(client);
            let conn = super::Connection::new(0, conn, config.clone());
            Driver::new(scenario, conn)
        };
        let mut client_trace = MemoryLogger::new(traces.clone());
        let mut client_checkpoints = HashSet::new();
        let mut client_timer = timer::Testing::default();

        let mut server = {
            let scenario = &scenario.servers[0].connections[0];
            let conn = Box::pin(server);
            let conn = super::Connection::new(1, conn, config);
            Driver::new(scenario, conn)
        };
        let mut server_trace = MemoryLogger::new(traces.clone());
        let mut server_checkpoints = HashSet::new();
        let mut server_timer = timer::Testing::default();

        let (waker, count) = new_count_waker();
        let mut prev_count = 0;
        let mut cx = core::task::Context::from_waker(&waker);

        loop {
            let c = client.poll_with_timer(
                &mut client_trace,
                &mut client_checkpoints,
                &mut client_timer,
                &mut cx,
            );
            let s = server.poll_with_timer(
                &mut server_trace,
                &mut server_checkpoints,
                &mut server_timer,
                &mut cx,
            );

            match (c, s) {
                (Poll::Ready(Ok(())), Poll::Ready(Ok(()))) => break,
                (Poll::Ready(Err(e)), _) | (_, Poll::Ready(Err(e))) => panic!("{}", e),
                _ => {
                    let current_count = count.get();
                    if current_count > prev_count {
                        prev_count = current_count;
                        continue;
                    }

                    if client_timer.advance_pair(&mut server_timer).is_none() {
                        eprintln!("the timer did not advance!");
                        eprintln!("server trace:");
                        eprintln!("{}", server_trace.as_str().unwrap());
                        eprintln!("{:#?}", server);
                        eprintln!("client trace:");
                        eprintln!("{}", client_trace.as_str().unwrap());
                        eprintln!("{:#?}", client);
                        panic!("test is deadlocked");
                    }
                }
            }
        }

        (client_trace, server_trace)
    }

    macro_rules! test {
        ($name:ident, $config:expr, $builder:expr) => {
            #[test]
            fn $name() -> crate::Result<()> {
                let scenario = Scenario::build(|scenario| {
                    let server = scenario.create_server();

                    scenario.create_client(|client| {
                        client.connect_to(server, $builder);
                    });
                });

                let (client_trace, server_trace) = test($config, &scenario);

                assert_display_snapshot!(
                    concat!(stringify!($name), "__client"),
                    client_trace.as_str().unwrap()
                );
                assert_display_snapshot!(
                    concat!(stringify!($name), "__server"),
                    server_trace.as_str().unwrap()
                );

                Ok(())
            }
        };
    }

    test!(single_stream, Config::default(), |conn| {
        conn.open_send_stream(
            |local| {
                local.send(1.kilobytes());
            },
            |remote| {
                remote.receive(1.kilobytes());
            },
        );
    });

    test!(single_slow_send_stream, Config::default(), |conn| {
        conn.open_send_stream(
            |local| {
                local.set_send_rate(100.bytes() / 50.millis());
                local.send(1.kilobytes());
            },
            |remote| {
                remote.receive(1.kilobytes());
            },
        );
    });

    test!(single_slow_recv_stream, Config::default(), |conn| {
        conn.open_send_stream(
            |local| {
                local.send(1.kilobytes());
            },
            |remote| {
                remote.set_receive_rate(100.bytes() / 50.millis());
                remote.receive(1.kilobytes());
            },
        );
    });

    test!(
        low_stream_window,
        Config {
            stream_window: 50,
            ..Config::default()
        },
        |conn| {
            conn.open_send_stream(
                |local| {
                    local.set_send_rate(100.bytes() / 50.millis());
                    local.send(1.kilobytes());
                },
                |remote| {
                    remote.receive(1.kilobytes());
                },
            );
        }
    );

    test!(
        low_max_streams,
        Config {
            max_streams: 2,
            peer_max_streams: 2,
            ..Config::default()
        },
        |conn| {
            conn.scope(|scope| {
                for _ in 0..4 {
                    scope.spawn(|conn| {
                        conn.open_send_stream(
                            |local| {
                                local.set_send_rate(500.bytes() / 50.millis());
                                local.send(1.kilobytes());
                            },
                            |remote| {
                                remote.receive(1.kilobytes());
                            },
                        );
                    });
                }
            });
        }
    );

    test!(multiple_streams, Config::default(), |conn| {
        conn.scope(|scope| {
            for _ in 0..3 {
                scope.spawn(|conn| {
                    conn.open_send_stream(
                        |local| {
                            local.set_send_rate(100.bytes() / 50.millis());
                            local.send(1.kilobytes());
                        },
                        |remote| {
                            remote.receive(1.kilobytes());
                        },
                    );
                });
            }
        });
    });

    test!(checkpoint, Config::default(), |conn| {
        let (park, unpark) = conn.checkpoint();
        conn.scope(|scope| {
            scope.spawn(|conn| {
                conn.open_send_stream(
                    |local| {
                        local.set_send_rate(100.bytes() / 50.millis());
                        local.send(1.kilobytes());
                        local.unpark(unpark);
                        local.send(1.kilobytes());
                    },
                    |remote| {
                        remote.receive(2.kilobytes());
                    },
                );
            });
            scope.spawn(|conn| {
                conn.open_send_stream(
                    |local| {
                        local.park(park);
                        local.set_send_rate(100.bytes() / 50.millis());
                        local.send(1.kilobytes());
                    },
                    |remote| {
                        remote.receive(1.kilobytes());
                    },
                );
            });
        });
    });
}
