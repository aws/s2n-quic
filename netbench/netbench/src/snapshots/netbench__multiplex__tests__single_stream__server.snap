---
source: netbench/src/multiplex.rs
assertion_line: 549
expression: (server_trace.1).as_str().unwrap()

---
0:00:00.000001: 1:accept: stream=0
0:00:00.000001: 1:1.0:exec: Receive { stream_id: 0, bytes: Byte(1000) }
0:00:00.000001: 1:1.0:recv: stream=0, len=1000
0:00:00.000001: 1:1.0:exec: ReceiveFinish { stream_id: 0 }
0:00:00.000001: 1:1.0:recv finish: stream=0

